core/src/test/kotlin/org/eln2/data/ComponentGraphTest.kt

```diff
@@ -0,0 +1,303 @@
+package org.eln2.data
+
+import org.eln2.debug.DEBUG
+import org.eln2.debug.dprintln
+import org.junit.jupiter.api.Assertions.*
+import org.junit.jupiter.api.Test
+
+class DebugListener: ComponentListener, VertexListener, EdgeListener {
+	override fun componentRealized(c: ConnectedComponent) {
+		dprintln("component realized: $c")
+	}
+
+	override fun componentUnrealized(c: ConnectedComponent) {
+		dprintln("component unrealized: $c")
+	}
+
+	override fun vertexAdded(v: Vertex) {
+		dprintln("vertex added: $v")
+	}
+
+	override fun vertexRemoved(v: Vertex) {
+		dprintln("vertex removed: $v")
+	}
+
+	override fun edgeAdded(e: Edge) {
+		dprintln("edge added: $e")
+	}
+
+	override fun edgeRemoved(e: Edge) {
+		dprintln("edge removed: $e")
+	}
+
+	companion object {
+		val instance = DebugListener()
+	}
+}
+
+internal class ComponentGraphTest {
+	fun makeInstance(): ComponentGraph = ComponentGraph().apply { addListener(DebugListener.instance) }
+	
+	fun debugDump(cg: ComponentGraph) {
+		dprintln("GRAPH $cg:")
+		dprintln("  vertices (${cg.vertices.size}):")
+		cg.vertices.forEach {
+			dprintln("    $it in ${it.component} with ${it.component.vertices}")
+			dprintln("      incidence ${it.incident} degree ${it.incident.size}")
+		}
+		dprintln("  edges (${cg.edges.size}):")
+		cg.edges.forEach {
+			dprintln("    $it between (${it.a}, ${it.b})")
+		}
+		dprintln("  components (${cg.components.size}):")
+		cg.components.forEach {
+			dprintln("    $it has ${it.vertices.size} vertices:")
+			it.vertices.forEach { v->
+				dprintln("      $v (component ${v.component})")
+			}
+		}
+	}
+    
+    fun fullConsistencyCheck(cg: ComponentGraph) {
+        try {
+            fullConsistencyCheckImpl(cg)
+        } catch(e: Throwable) {
+            DEBUG = true
+            System.err.println("The following graph failed consistency checks:")
+            debugDump(cg)
+            throw e
+        }
```

Comment:
This should be an assert plus a debug print, rather than a throw and an error print.

Reply:
The `try/catch` here actually catches assertions and rethrows them; it's `fullConsistencyCheckImpl`'s job to raise an assertion whenever it finds an inconsistency. The problem is that the assertion error doesn't have the nice full graph, so I conditionally dump the _entire_ graph whenever a consistency check fails. Due to the reliance on the "old" debug architecture and tons of `dprintln`, I forced debugging on globally as a bit of a last-minute dirty hack to ensure that the graph _would_ make it to the log.

This infrastructure has been _extremely_ helpful for debugging issues with the data model, but it definitely still is a hack. I haven't looked at the new debug system, but I'm hoping that would fix this :)

Reply:
The point of using `dprintln()` is to get it into the signature. We should be using it everywhere.

The reason I want an assert is because I *don't* want a re-throw. If you need more details, print the information you want to see before calling `assert()`. This will end the test early, and show the error promptly.

Reply:
I'll let this slide for now.